Receive mesh rule

orlandoisepic opened the pull request:
Update 25.5.25:
New in this version is a rule that checks for unused receive meshes!
- A receive-mesh is unused, if a participant receives it (without api-access), but does not map it to their own mesh. 
- It is also unused, if a participant receives it with api-access, but does not operate directly on it.
- Bonus violation: If a participant receives a mesh with api-access, they do not have to map it. If they still do, they get a debug-violation!

Have fun enjoying this new functionality!

Closes #7

orlandoisepic committed:
Create receive_mesh.py

A new rule to avoid unused receive-meshes!

orlandoisepic committed:
Create receive-mesh_test.py

a test for receive mesh rule

orlandoisepic committed:
Update receive-mesh_test.py

orlandoisepic committed:
Create precice-config.xml

config for test

orlandoisepic committed:
Adjust tests to receive-mesh rule

It seems like we have quite some unused receive meshes

orlandoisepic committed:
Update rules_processing.py

Add new rule

TheFirstGameruler commented on preciceconfigchecker/rules/receive_mesh.py:
```diff
@@ -0,0 +1,122 @@
+                            if mapping.to_mesh == mesh or mapping.from_mesh == mesh:
+                                used = True
+                                violations.append(self.MappedAPIAccessReceiveMesh(participant, mesh))
+                        # if mapping.direction == Direction.READ:
+                        #     # In a regular mapping, the to-mesh must not be None
+                        #     # If it is None, then the mapping is a just-in-time mapping
+                        #     if mapping.from_mesh == mesh and mapping.to_mesh is not None:
+                        #         used = True
+                        #         violations.append(self.MappedAPIAccessReceiveMesh(participant, mesh))
+                        # elif mapping.direction == Direction.WRITE:
+                        #     if mapping.to_mesh == mesh and mapping.from_mesh is not None:
+                        #         used = True
+                        #         violations.append(self.MappedAPIAccessReceiveMesh(participant, mesh))
```
forgot to remove?

orlandoisepic replied:
yup, thanks

TheFirstGameruler commented on preciceconfigchecker/rules/receive_mesh.py:
```diff
@@ -0,0 +1,122 @@
+                    used = True
+                # These are only valid if they have api-access:
+                if receive_mesh.api_access:
+                    for read_data in participant.read_data:
+                        if read_data.mesh == mesh:
+                            used = True
```
Theoretically, you can use `break` if the condition is true to break the loop. Multiple times possible.

orlandoisepic replied:
True. Thanks

TheFirstGameruler replied:
Is the `if used or read_data.mesh == mesh` part too much?

TheFirstGameruler reviewed (dismissed):
Top.
Just minor changes that wouldn't be that important.

orlandoisepic committed:
Update Rules.md

updated documentation. changed some explanations of TODO rules and improved wording.
added explanation of receive mesh rule

orlandoisepic committed:
Update receive_mesh.py

improve checks to do less when it is not needed

orlandoisepic committed:
Update receive_mesh.py

slightly adjust check

TheFirstGameruler reviewed (approved):
Top
